src/test/java/org/cactoos/iterator/SyncIteratorTest.java

```diff
@@ -73,4 +78,209 @@ public void syncIteratorReturnsCorrectValuesWithInternalLock() {
         );
     }
 
+    @Test
+    @SuppressWarnings("PMD.AvoidInstantiatingObjectsInLoops")
+    public void correctValuesForConcurrentNextNext()
+        throws InterruptedException {
+        for (int iter = 0; iter < 5000; iter += 1) {
+            final List<String> list = Arrays.asList("a", "b");
+            final SyncIterator<String> iterator = new SyncIterator<>(
+                list.iterator()
+            );
+            final List<Object> sync =
+                Collections.synchronizedList(
+                    new ArrayList<>(list.size())
+                );
+            final Runnable first = () -> {
+                sync.add(iterator.next());
+            };
+            final Runnable second = () -> {
+                sync.add(iterator.next());
+            };
+            final List<Runnable> test = new ArrayList<>(list.size());
+            test.add(first);
+            test.add(second);
+            new Concurrent(test).launch();
+            MatcherAssert.assertThat(
+                "Missing the list items(s) (next()).",
+                sync,
+                Matchers.containsInAnyOrder("a", "b")
+            );
+        }
+    }
+
+    @Test
+    @SuppressWarnings("PMD.AvoidInstantiatingObjectsInLoops")
+    public void correctValuesForConcurrentNextHasNext()
+        throws InterruptedException {
+        for (int iter = 0; iter < 5000; iter += 1) {
+            final List<String> list = Arrays.asList("a", "b");
+            final SyncIterator<String> iterator = new SyncIterator<>(
+                list.iterator()
+            );
+            final List<Object> sync =
+                Collections.synchronizedList(
+                    new ArrayList<>(list.size())
+                );
+            final Runnable first = () -> {
+                sync.add(iterator.next());
+            };
+            final Runnable second = () -> {
+                sync.add(iterator.next());
+            };
+            final Runnable third = () -> {
+                sync.add(iterator.hasNext());
+            };
+            final List<Runnable> test = new ArrayList<>(list.size() + 1);
+            test.add(first);
+            test.add(second);
+            test.add(third);
+            new Concurrent(test).launch();
+            MatcherAssert.assertThat(
+                "Missing the list items(s) (next()).",
+                sync,
+                Matchers.allOf(
+                    Matchers.hasItem("a"),
+                    Matchers.hasItem("b")
+                )
+            );
+            MatcherAssert.assertThat(
```

Comment:
Two matchers have to be used here, because the compiler complains about using `Object` arrays in matchers (contaisInAnyOrder with different object types). (produces a compiler warning)